src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -607,6 +656,41 @@ var ReactCompositeComponentMixin = {
       inst.props = nextProps;
       inst.state = nextState;
       inst.context = nextContext;
+
+      // We will also, potentially, want to update components down the tree
+      // if the context has updated
+      var shouldUpdateChildContext = !inst.shouldUpdateChildContext ||
+        inst.shouldUpdateChildContext(nextProps, nextState, nextContext);
+
+      if (shouldUpdateChildContext) {
```

Comment:
Developers will be tempted to just always return true for shouldUpdateChildContext, since it's the simplest thing to do.  At the very least, we should warn if they return true but the object that getChildContext() returns is shallow equal to the previous context they returned.  We might even want to make `shouldUpdateChildContext` mandatory for any context provider.  Thoughts @sebmarkbage?